students/charlie_england/session07/test_html_render.py

```diff
@@ -0,0 +1,349 @@
+    assert file_contents.startswith("<body>")
+    assert file_contents.endswith("</body>")
+
+
+def test_p():
+    e = P("this is some text")
+    e.append("and this is some more text")
+
+    file_contents = render_result(e).strip()
+
+    assert("this is some text") in file_contents
+    assert("and this is some more text") in file_contents
+
+    assert file_contents.startswith("<p>")
+    assert file_contents.endswith("</p>")
+
+
+def test_sub_element():
+    """
+    tests that you can add another element and still render properly
+    """
+    page = Html()
+    page.append("some plain text.")
+    page.append(P("A simple paragraph of text"))
+    page.append("Some more plain text.")
+
+    file_contents = render_result(page)
+    print(file_contents) # so we can see it if the test fails
+
+    # note: The previous tests should make sure that the tags are getting
+    #       properly rendered, so we don't need to test that here.
+    assert "some plain text" in file_contents
+    assert "A simple paragraph of text" in file_contents
+    assert "Some more plain text." in file_contents
+    assert "some plain text" in file_contents
+    # but make sure the embedded element's tags get rendered!
+    assert "<p>" in file_contents
+    assert "</p>" in file_contents
+
+
+
+
+########
+# Step 3
+########
+
+def test_head():
+    hd = Head("head text here")
+    hd.append("more head text here")
+
+    file_contents = render_result(hd).strip()
+
+    assert "head text here" in file_contents
+    assert "more head text here" in file_contents
+
+    #testing subelement
+
+    page = Html()
+    page.append("this is the html")
+    page.append(Head("and this is the head"))
+    file_contents = render_result(page)
+    assert "this is the html" in file_contents
+    assert "and this is the head" in file_contents
+    # but make sure the embedded element's tags get rendered!
+    assert "<head>" in file_contents
+    assert "</head>" in file_contents
+
+def test_onelinetag():
+    olt = OneLineTag("hello, its me")
+    olt.append("calling from the same line")
+
+    file_contents = render_result(olt).strip()
+
+    assert "hello, its me" in file_contents
+    assert "calling from the same line" in file_contents
+
+def test_title():
+    e = Title("Here is some title text")
+
+    file_contents = render_result(e)
+
+    assert "Here is some title text" in file_contents
+    print(file_contents)
+    assert file_contents.startswith("<title>")
+
+def test_ele_attributes():
+    attributes = {"id":"TheList","style":"line-height:200%"}
+    e = Body("here is some body text",**attributes)
+
+    file_contents = render_result(e)
+    assert "here is some body text" in file_contents
+    assert 'id="TheList"' in file_contents
+
+#I had a lot of trouble with this part and did not understand the tutorial:
+#If you put the code to render the opening (and closing) tags in it’s own method, then we can call that method from multiple render methods, something like
```

Comment:
in your case, you rendered the opening tag right away when the element was created -- so yes, this doesn't make sense in that context.

Better to render to create the tag at render time -- then you can have a method that just does that. And override it when you want to do it differently.